plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java

```diff
@@ -0,0 +1,382 @@
+    {
+        // try to use result set metadata from select * from table to populate the mapping
+        try {
+            HashMap<String, CaseSensitivity> caseMap = new HashMap<>();
+            String sql = format("select * from %s.%s where 0=1", schemaTableName.getSchemaName(), schemaTableName.getTableName());
+            PreparedStatement pstmt = connection.prepareStatement(sql);
+            ResultSetMetaData rsmd = pstmt.getMetaData();
+            int columnCount = rsmd.getColumnCount();
+            for (int i = 1; i <= columnCount; i++) {
+                caseMap.put(rsmd.getColumnName(i), rsmd.isCaseSensitive(i) ? CASE_SENSITIVE : CASE_INSENSITIVE);
+            }
+            pstmt.close();
+            return caseMap;
+        }
+        catch (SQLException e) {
+            // behavior of base jdbc
+            return ImmutableMap.of();
+        }
+    }
+
+    @Override
+    public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    {
+        // this method should ultimately encompass all the expected teradata data types
+        Optional<ColumnMapping> mapping = getForcedMappingToVarchar(typeHandle);
+        if (mapping.isPresent()) {
+            return mapping;
+        }
+
+        switch (typeHandle.jdbcType()) {
+            case Types.TINYINT:
+                return Optional.of(tinyintColumnMapping());
+            case Types.SMALLINT:
+                return Optional.of(smallintColumnMapping());
+            case Types.INTEGER:
+                return Optional.of(integerColumnMapping());
+            case Types.BIGINT:
+                return Optional.of(bigintColumnMapping());
+            case Types.REAL:
+            case Types.DOUBLE:
+            case Types.FLOAT:
+                // teradata float is 64 bit
+                // trino double is 64 bit
+                // teradata float / real / double precision all map to jdbc type float
+                return Optional.of(doubleColumnMapping());
+            case Types.NUMERIC:
+            case Types.DECIMAL:
+                return numberMapping(typeHandle);
+            case Types.CHAR:
+                return Optional.of(charColumnMapping(typeHandle.requiredColumnSize(), deriveCaseSensitivity(typeHandle.caseSensitivity().orElse(null))));
+            case Types.VARCHAR:
+                // see prior note on trino case sensitivity
+                return Optional.of(varcharColumnMapping(typeHandle.requiredColumnSize(), deriveCaseSensitivity(typeHandle.caseSensitivity().orElse(null))));
+            case Types.DATE:
+                return Optional.of(dateColumnMappingUsingLocalDate());
+        }
+
+        if (getUnsupportedTypeHandling(session) == CONVERT_TO_VARCHAR) {
+            return mapToUnboundedVarchar(typeHandle);
+        }
+
+        return Optional.empty();
+    }
+
+    private Optional<ColumnMapping> numberMapping(JdbcTypeHandle typeHandle)
```

Comment:
This method can be static.

Reply:
changed. Modified other mapping methods - charColumnMapping, varcharColumnMapping